### PR TITLE
Harden redirect handler: buffer content, filter sensitive headers

### PR DESCRIPTION
**Pull request by Chris0Jeky**
## Summary
- **Content buffering**: Read request body into byte array before first `SendAsync` so 307/308 redirects create fresh `ByteArrayContent` instead of reusing a potentially consumed/non-seekable stream
- **Host header stripping**: Always excluded on redirects since it must match the new destination
- **Authorization header filtering**: Stripped on cross-host redirects to prevent credential leakage; preserved on same-host redirects

Addresses Gemini HIGH findings on PR #1054.

## Test plan
- [x] `SendAsync_307Redirect_PreservesContentAndSafeHeaders` — verifies body is replayed and custom headers forwarded
- [x] `SendAsync_307CrossHostRedirect_StripsAuthorizationHeader` — verifies auth stripped on cross-host
- [x] `SendAsync_307SameHostRedirect_PreservesAuthorizationHeader` — verifies auth kept on same-host
- [x] All 14 EgressEnvelopeHandler tests pass

### CODE REVIEW

File: backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs [hunk 1]
```diff
@@ -42,6 +42,18 @@ protected override async Task<HttpResponseMessage> SendAsync(
         // against the egress allowlist. If auto-redirect is enabled, the handler only
         // sees the final response and the redirect check becomes ineffective.
 
+        // Buffer request content before the first send so it can be replayed on 307/308 redirects.
+        // After SendAsync, the original content stream may be consumed and non-seekable.
+        byte[]? bufferedContent = null;
+        string? contentType = null;
+        if (request.Content is not null)
+        {
+            bufferedContent = await request.Content.ReadAsByteArrayAsync(cancellationToken);
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs]: Fixed in `4b5a4420`. Redirect replay is now capped by `MaxRedirectReplayContentBytes` (1 MiB); unknown-length or oversized content is not eagerly buffered and fails closed if a 307/308 replay is required. Added regression coverage for unknown-length no-redirect streaming and 307 oversized/unknown-length fail-closed behavior. Verified with `dotnet test backend/tests/Taskdeck.Application.Tests/Taskdeck.Application.Tests.csproj -c Release --filter "FullyQualifiedName~EgressEnvelopeHandlerTests"` (21 passed).

File: backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs [hunk 2]
```diff
@@ -81,12 +93,24 @@ protected override async Task<HttpResponseMessage> SendAsync(
                 Method = statusCode is 307 or 308 ? request.Method : HttpMethod.Get,
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs]: Fixed in `4b5a4420`. The redirect loop now tracks the current request per hop and uses the previous/current hop method for 307/308 instead of reverting to the original method. Added `302 -> 307` coverage proving the final request remains `GET`. Verification: `EgressEnvelopeHandlerTests` (21 passed).

File: backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs [hunk 3]
```diff
@@ -81,12 +93,24 @@ protected override async Task<HttpResponseMessage> SendAsync(
...
-            // 307/308 require preserving the original headers and body
+            // 307/308 require preserving the original body and safe headers
             if (statusCode is 307 or 308)
             {
-                redirectRequest.Content = request.Content;
+                if (bufferedContent is not null)
+                {
+                    var newContent = new ByteArrayContent(bufferedContent);
+                    if (contentType is not null)
+                        newContent.Headers.TryAddWithoutValidation("Content-Type", contentType);
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs]: Fixed in `4b5a4420`. Replayable content now captures and reapplies all original content headers except `Content-Length`, including `Content-Encoding`, `Content-Language`, and custom content headers. Added assertions for those headers in the 307 replay test. Verification: `EgressEnvelopeHandlerTests` (21 passed).

File: backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs [hunk 4]
```diff
@@ -81,12 +93,24 @@ protected override async Task<HttpResponseMessage> SendAsync(
...
+                    redirectRequest.Content = newContent;
+                }
+
+                var isCrossHost = !string.Equals(originalHost, resolvedRedirectUri.Host, StringComparison.OrdinalIgnoreCase);
                 foreach (var header in request.Headers)
                 {
+                    if (string.Equals(header.Key, "Host", StringComparison.OrdinalIgnoreCase))
+                        continue;
+                    if (isCrossHost && string.Equals(header.Key, "Authorization", StringComparison.OrdinalIgnoreCase))
+                        continue;
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs]: Fixed in `4b5a4420`. Cross-origin redirects now strip `Cookie` plus other credential-like headers (`Authorization`, `Proxy-Authorization`, API-key headers, token/secret headers). Origin comparison uses scheme + host + port, so same-host scheme/port changes also strip credentials. Added regression coverage for cookie/API-key stripping and same-host scheme/port changes. Verification: `EgressEnvelopeHandlerTests` (21 passed).

File: backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs [hunk 5]
```diff
@@ -81,12 +93,24 @@ protected override async Task<HttpResponseMessage> SendAsync(
...
+            // 307/308 require preserving the original body and safe headers
             if (statusCode is 307 or 308)
             {
-                redirectRequest.Content = request.Content;
+                if (bufferedContent is not null)
+                {
+                    var newContent = new ByteArrayContent(bufferedContent);
+                    if (contentType is not null)
+                        newContent.Headers.TryAddWithoutValidation("Content-Type", contentType);
+                    redirectRequest.Content = newContent;
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs]: Fixed in `4b5a4420`. The replay path now captures the full content-header set before the first send and reapplies it on each 307/308 replay. Test coverage includes `Content-Encoding`, `Content-Language`, and a custom content signature header. Verification: `EgressEnvelopeHandlerTests` (21 passed).

File: backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs [hunk 6]
```diff
@@ -42,6 +42,18 @@ protected override async Task<HttpResponseMessage> SendAsync(
...
         // sees the final response and the redirect check becomes ineffective.
 
+        // Buffer request content before the first send so it can be replayed on 307/308 redirects.
+        // After SendAsync, the original content stream may be consumed and non-seekable.
+        byte[]? bufferedContent = null;
+        string? contentType = null;
+        if (request.Content is not null)
+        {
+            bufferedContent = await request.Content.ReadAsByteArrayAsync(cancellationToken);
+            contentType = request.Content.Headers.ContentType?.ToString();
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs]: Fixed in `4b5a4420`. Unknown-length and oversized bodies are no longer eagerly read on the normal no-redirect path; they stream through unchanged. If a 307/308 redirect later requires replay and the body is not replayable within the 1 MiB cap, the handler fails closed instead of buffering unbounded content. Verification: `EgressEnvelopeHandlerTests` (21 passed).

File: backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs [hunk 7]
```diff
@@ -139,6 +165,67 @@ private static bool IsRedirect(HttpResponseMessage response)
...
+
+        var content = await request.Content.ReadAsByteArrayAsync(cancellationToken);
+        if (content.LongLength > MaxRedirectReplayContentBytes)
+        {
+            throw new InvalidOperationException(
+                $"Request content exceeds the {MaxRedirectReplayContentBytes} byte redirect replay limit.");
+        }
+
+        var replayContent = new ReplayableContent(content, headers);
+        request.Content = CreateReplayContent(replayContent);
```
[the remark this answers is not in the document]
Comment (Chris0Jeky) [reply backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs:189]: Fixed in `fbc336b7`. After buffering bounded replay content, the handler swaps in a new `ByteArrayContent` and disposes the original content instance so wrapped streams/handles are not orphaned. Added a `TrackingContent` regression. Verification: `EgressEnvelopeHandlerTests` now 22 passed.

**Comment by Chris0Jeky** [comment]
## Security Review Findings

### CRITICAL
None.

### HIGH
- `backend/src/Taskdeck.Application/Services/EgressEnvelopeHandler.cs:185` - `PrepareReplayableContentAsync` trusts `Content-Length` before calling `ReadAsByteArrayAsync`, but a custom or misbehaving `HttpContent` can report a length below `MaxRedirectReplayContentBytes` and still serialize far more data. `ReadAsByteArrayAsync` buffers the whole body before the later `content.LongLength` check runs, so the redirect replay cap can still be bypassed and can exhaust memory before the handler fails closed. Expected fix: enforce the cap while reading, for example by copying `ReadAsStreamAsync(cancellationToken)` into a bounded buffer in chunks and throwing as soon as the running total exceeds `MaxRedirectReplayContentBytes` (or by using a max-buffer API that fails before allocating the oversized payload). Add a regression with a custom `HttpContent` whose `Headers.ContentLength` is small but whose serialized body exceeds the cap, and assert the inner handler is not invoked and the handler fails before buffering beyond the limit.

### MEDIUM
None.

### LOW
None.

I inspected the PR body, commits, full diff, checks, reviews, issue comments, and review threads. Existing bot findings are replied to and resolved; this is a remaining head-commit issue in the redirect replay hardening path.

**Comment by Chris0Jeky** [comment]
## Review Finding Fix Evidence

Addressed the HIGH finding from https://github.com/Chris0Jeky/Taskdeck/pull/1055#issuecomment-4446180858.

Fix commit: c9d050ae0add9380659c467bcbbb9425a8d412eb

What changed:
- Replaced redirect replay buffering via `ReadAsByteArrayAsync` with `CopyToAsync` into a bounded replay buffer, so serialization fails as soon as the redirect replay cap would be exceeded.
- Added a regression where `HttpContent.Headers.ContentLength` reports `1` byte while serialization attempts `MaxRedirectReplayContentBytes + 1`; the test asserts the inner handler is not invoked and the extra write does not complete.

Verification:
- `powershell -File scripts/check-git-env.ps1` - passed.
- `dotnet test backend/tests/Taskdeck.Application.Tests/Taskdeck.Application.Tests.csproj -c Release --filter FullyQualifiedName~EgressEnvelopeHandlerTests` - passed: 23/23 tests.

Docs:
- No docs updated; this is a narrow PR review fix with no shipped-product or roadmap documentation change.

**Comment by Chris0Jeky** [comment]
## Second Adversarial Review -- No New Findings

**Reviewer**: Claude (Opus 4.6)  
**Scope**: Full diff (4 commits), all existing comments (gemini-code-assist, chatgpt-codex-connector, first adversarial review + fix evidence), head-commit source + tests.

### Analysis Summary

| Area | Verdict |
|------|---------|
| Bounded buffering (`BoundedReplayBuffer`) | Correct. `ThrowIfExceedsLimit` enforces the 1 MiB cap per-write before bytes reach the inner `MemoryStream`. |
| Header safety (`IsSensitiveRedirectHeader`) | Comprehensive. Covers Authorization, Proxy-Authorization, Cookie, known API-key headers, and pattern-matches `token`/`secret`. Fail-safe (over-strip rather than under-strip). |
| Same-origin comparison (`IsSameOrigin`) | Strict triple check (scheme + host + port). `Uri.Port` returns the default port for the scheme, so implicit vs. explicit port comparison is correct. |
| Content replay lifecycle | Original content is disposed after buffering. `ByteArrayContent` replays include all captured content headers except `Content-Length` (which .NET sets automatically). |
| Fail-closed for non-replayable content | Correctly disposes the redirect response before throwing `InvalidOperationException`. |
| Method preservation across hops | Uses `previousRequest.Method` (not the original), so a 302-then-307 correctly stays GET. Tested. |
| Relative URI resolution | Resolves against `currentRequest.RequestUri`, not the original. Tested. |
| `replayContent = null` on non-307/308 | Frees the buffered byte[] for GC; subsequent 307/308 hops won't attempt replay since the intermediate request has no content. |
| Test coverage | 23 tests covering content preservation, credential stripping (6 header types), same-origin edge cases (scheme, port), chained redirects, unknown-length streaming, oversized, under-reported content, and disposal. |
| CI | All checks green (Backend Unit x2, API Integration x2, Frontend Unit x2, E2E Smoke, Architecture, Migration, Docs Governance, CodeQL x4, Secrets Detection, Workflow Lint, Container Images, Paper Color Audit, OpenAPI Guardrail, Dependency Review). |

### Previous Findings Status

All findings from round 1 (HIGH: unbounded buffering bypass via under-reported Content-Length) and bot reviews (gemini-code-assist: 4 findings; chatgpt-codex-connector: 2 findings) have been addressed with fix commits and verification evidence posted inline.

### Pre-existing Note (Out of Scope)

The redirect response is not disposed before throwing `EgressViolationException` on egress violations within the redirect loop. This pattern existed prior to this PR on `main` and is not introduced by these changes. Would be suitable for a follow-up issue if desired.

### Conclusion

**No findings. All previous findings addressed. CI green. Ready for merge consideration.**
